### PR TITLE
Change `done` field to integers for handling truncation vs termination

### LINKED ISSUE #2510: [Proposal] Formal API handling of truncation vs termination
Right now, one of the biggest weaknesses of the Gym API is that Done is used for both truncation and termination. The problem is that algorithms in Q learning family (and I assume others), depend on the differentiation between a terminal state and a truncated state to function as intended. Right now, Gym solves this for some (but not all) environments with infos for built in environments via the time limiting wrapper inserting this information into infos. No third party environments I'm aware of use this, and most learning code uses this. This has resulted in a few problems:

1. The DQN people are deeply unamused because proper reimplementation of their algorithms are very frequently done wrong
2. Many people who are very qualified in RL now have complained about being wildly confused by this while first learning RL
3. Many large institutions (e.g. some large groups at Google brain) refuse to use Gym almost entirely over this design issue, which is bad

This sort of thing in the opinion of myself and those I've spoken to at OpenAI warrants a breaking change in the pursuit of a 1.0 release. There are three options for making the breaking change:

1. Add have step return an extra boolean value in addition to done, e.g. something like `observation, reward, done/terminated, truncated, info = env.step(action)`
2. That option, but return the discount value like dm_env does instead of a boolean
3. Turn `done` into a custom three state variable, or give it a method like `done.is_truncated`

My preference is option 1. I think that option 3 is too confusing and wouldn't generally be used because it's not explicit. I also think that option 2 adds additional confusion and complexity over option 1 without really adding any additional utility. I'm creating this issue so that people can discuss it before a PR is made. This feature is intended to be released in the release of gym after the current.

Regarding breaking changes and backwards compatibility:

- Old style environments can just be wrapped to add this, and we'll include a function in gym.wrappers to do this
- Learning code can be compliant with the new API simply via an extra `_,`

In other words, it's my view that the level of breaking this update will be is a minor annoyance, and I think that this is a very important feature. This change will be the second of three planned small changes to the base API before the 1.0 release (the first was to the seed method, the last will be to the render method). I am obviously aware of the profound role Gym plays in the community, however Gym will almost certainly be used a lot more over the next 5 years than the past 5, and I think that a series of minor breaking changes to clean things up now in the pursuit of a proper 1.0 release will be very worth it.

Please offer your thoughts on which of the three options here is best in the comments if you would like (or new options if I'm missing something) and remain civil and inviting to others.

### PR DESCRIPTION
**Pull request by XuehaiPan**
Change value range for `done` from `(False, True)` to `(0, 1, 2)`. See https://github.com/openai/gym/issues/2510#issuecomment-1035960224 for more details.

Resolves #2510

------

#### Type and Value Range

```
done: bool
  - False: not done
  - True:  terminated or truncated
```

#### Sampling

```python
observation = env.reset()
done = False
while not done:
    action = policy(observation)
    observation, reward, done, info = env.step(action)
```

#### Value iteration

```python
V_target = reward + gamma * (1 - done) * V(next_observation)
```

```python
1 - done  # -> 1  (not done)
1 - done  # -> 0  (terminated or truncated)
```

This cannot distinguish **truncation** and **termination**. The `done` value for both situation is `True`.

------

#### Type and Value Range

```
done: int
  - gym.Env.NOT_DONE   (0): not done
  - gym.Env.TERMINATED (1): terminated (internal done in environment)
  - gym.Env.TRUNCATED  (2): truncated  (truncated by wrappers or step counting)
```

NOTE: It's better to reference constants by name rather than value.

```python
done = env.NOT_DONE  # :)
done = 0  # :(
```

#### Sampling

```python
observation = env.reset()
done = False
while not bool(done):
    action = policy(observation)
    observation, reward, done, info = env.step(action)
```

NOTE: in `if expression` / `while experssion` / `not expression`, the explicit type casting `bool(done)` can be ommited.

#### Value iteration

```python
V_target = reward + gamma * abs(1 - done) * V(next_observation)
```

The only change is:

```python
(1 - done) --> abs(1 - done)

abs(1 - done)  # -> 1  (not done or truncated)
abs(1 - done)  # -> 0  (terminated)
```

or

```python
(1 - done) --> int(done != env.TERMINATED)

int(done != env.TERMINATED)  # -> 1  (not done or truncated)
int(done != env.TERMINATED)  # -> 0  (terminated)
```

### CODE REVIEW

File: gym/vector/async_vector_env.py [hunk 1]
```diff
@@ -648,7 +648,7 @@ def _worker(index, env_fn, pipe, parent_pipe, shared_memory, error_queue):
 
             elif command == "step":
                 observation, reward, done, info = env.step(data)
-                if done:
+                if bool(done):
```
Comment (RedTachyon) [inline gym/vector/async_vector_env.py:651]: I don't think this actually does anything, the conditional works even with an integer:
```python
if 2:
    print("Hi")
> Hi
```

An argument might be made for readability, but I don't think casting to bool actually makes it more readable

Comment (RedTachyon) [reply gym/vector/async_vector_env.py:651]: The fact that `if done` still works is actually a pretty good showcase of the (mostly) backwards-compatibility of this

Comment (XuehaiPan) [reply gym/vector/async_vector_env.py:651]: > An argument might be made for readability, but I don't think casting to bool actually makes it more readable

Yes, I put this for readability. This explicit type casting is made for _"keep in mind that `expression` is not a boolean in statement `bool(expression)`"_.

**Comment by jkterry1** [comment]
@XuehaiPan could you please fix the merge conflicts?

**Comment by XuehaiPan** [comment]
Closed because it is preferred to explicitly add two booleans `terminated` and `truncated`.

**Pull request closed without merging**